src/DotNetty.Handlers/Tls/SniHandler.cs

```diff
@@ -0,0 +1,287 @@
+#endif
+                                                this.Select(context, hostname);
+                                                return;
+                                            }
+                                            else
+                                            {
+                                                // invalid enum value
+                                                continueLoop = false;
+                                                break;
+                                            }
+                                        }
+
+                                        offset += extensionLength;
+                                    }
+                                }
+
+                                break;
+                            // Fall-through
+                            default:
+                                //not tls, ssl or application data, do not try sni
+                                continueLoop = false;
+                                break;
+                        }
+                    }
+                }
+                catch (Exception e)
+                {
+                    error = e;
+
+                    // unexpected encoding, ignore sni and use default
+                    if (Logger.DebugEnabled)
+                    {
+                        Logger.Warn($"Unexpected client hello packet: {ByteBufferUtil.HexDump(input)}", e);
+                    }
+                }
+
+                if (this.serverTlsSniSettings.DefaultServerHostName != null)
+                {
+                    // Just select the default certifcate
+                    this.Select(context, this.serverTlsSniSettings.DefaultServerHostName); 
+                }
+                else
+                {
+                    this.handshakeFailed = true;
+                    var e = new DecoderException($"failed to get the Tls Certificate {error}");
+                    TlsUtils.NotifyHandshakeFailure(context, e);
+                    throw e;
+                }
+            }
+        }
+
+        async void Select(IChannelHandlerContext context, string hostName)
+        {
+            Contract.Requires(hostName != null);
+            try
+            {
+                this.ReplaceHandler(context, await this.serverTlsSniSettings.ServerCertificateSelector(hostName));
+            }
+            catch (Exception ex)
+            {
+                this.ExceptionCaught(context, new DecoderException($"failed to get the Tls Certificate for {hostName}, {ex}"));
+            }
+        }
+
+        void ReplaceHandler(IChannelHandlerContext context, X509Certificate2 tlsCertificate)
+        {
+            Contract.Requires(tlsCertificate != null);
+            var serverTlsSetting = new ServerTlsSettings(tlsCertificate, this.serverTlsSniSettings.NegotiateClientCertificate, this.serverTlsSniSettings.CheckCertificateRevocation, this.serverTlsSniSettings.EnabledProtocols);
```

Comment:
looking at the initialization pattern here and in tests it feels a bit too constrained to use the same settings for different endpoints. E.g. if you're co-hosting endpoints where one would negotiate client certs while the other wouldn't, or where you would allow older protocols on a separate endpoint. What do you think about changing the map to (string -> Task<ServerTlsSettings)?